# MC316 Practice Test – PYTHON

Please read the questions carefully. You will have 30 minutes to finish the test. 

#### Please import pandas library for use in scripting in the cell below. 


In [3]:
import pandas as pd

#### Please import the datasets and name them. 
For this test, you will use employee compensation data provided by Kansas Department of Administration. Download the original data here: <http://kanview.ks.gov/DataDownload.aspx>. 
For your convenience, the datasets were downloaded and uploaded to the instructor's database. You can retrieve the url for the first dataset from  <https://datajournalism.tech/wp-content/uploads/2019/05/EmployeeData_2017.csv> and the second dataset from
<https://datajournalism.tech/wp-content/uploads/2019/05/EmployeeData_2018.csv>

In [24]:
first = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/EmployeeData_2017.csv")

In [25]:
second = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/EmployeeData_2018.csv")

#### Please describe the two datasets.

In [26]:
first.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39422 entries, 0 to 39421
Data columns (total 7 columns):
Agency_Number      39422 non-null int64
Agency_Name        39422 non-null object
Employee_Name      39422 non-null object
Job_Title          39422 non-null object
Total_Gross_Pay    39422 non-null float64
Overtime_Pay       8615 non-null float64
Unnamed: 6         0 non-null float64
dtypes: float64(3), int64(1), object(3)
memory usage: 2.1+ MB


In [27]:
second.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39299 entries, 0 to 39298
Data columns (total 8 columns):
Fiscal_Year        39299 non-null int64
Agency_Number      39299 non-null int64
Agency_Name        39299 non-null object
Employee_Name      39299 non-null object
Job_Title          39299 non-null object
Total_Gross_Pay    39299 non-null float64
Overtime_Pay       39299 non-null float64
Unnamed: 7         0 non-null float64
dtypes: float64(3), int64(2), object(3)
memory usage: 2.4+ MB


#### Question 1: In the year of 2017,  which agency has the most employees? 
After writing the code, make sure you give the answer in the cell right below the code.

In [28]:
first.Agency_Name.value_counts()

Kansas State University                5367
University of Kansas                   5207
University of Kansas Medical Center    3419
Wichita State University               2248
Department of Transportation           2180
Dept for Children and Families         2130
Judicial Branch                        1833
Fort Hays State University             1411
Department of Revenue                  1033
Dept of Health & Environment           1010
Pittsburg State University              957
KS Wildlife  Parks & Tourism            815
Highway Patrol                          787
Emporia State University                761
Larned State Hospital                   669
Lansing Correctional Facility           540
Department of Administration            448
Hutchinson Correctional Fclty           446
Parsons St Hosp & Trng Center           433
Osawatomie State Hospital               390
Kansas Neurological Institute           389
Legislature                             383
Department of Corrections       

Answer: Kansas State University                5367

#### Question 2: In the year of 2018, does this agency still have the most employees? If yes, is there any change in the total number of employees? If no, what's the name of the agency that had the most employees?

In [30]:
second.Agency_Name.value_counts().head(5)

Kansas State University                5464
University of Kansas                   5281
University of Kansas Medical Center    3622
Wichita State University               2359
Department of Transportation           2104
Name: Agency_Name, dtype: int64

In [31]:
5464-5367

97

Answer:97

#### Question 3: How many employees that Kansas State University had in the year of 2018? Who are these people and how much they got paid?
Provide a subset of data and name it `KSU`. Describe the subset.

In [36]:
KSU = second[second.Agency_Name == "Kansas State University"]

In [37]:
KSU

Fiscal_Year  Agency_Number              Agency_Name  \
10323         2018            367  Kansas State University   
10324         2018            367  Kansas State University   
10325         2018            367  Kansas State University   
10326         2018            367  Kansas State University   
10327         2018            367  Kansas State University   
10328         2018            367  Kansas State University   
10329         2018            367  Kansas State University   
10330         2018            367  Kansas State University   
10331         2018            367  Kansas State University   
10332         2018            367  Kansas State University   
10333         2018            367  Kansas State University   
10334         2018            367  Kansas State University   
10335         2018            367  Kansas State University   
10336         2018            367  Kansas State University   
10337         2018            367  Kansas State University   
10338         2018            367  Kansas State University   
10339         2018            367  Kansas State University   
10340         2018            367  Kansas State University   
10341         2018            367  Kansas State University   
10342         2018            367  Kansas State University   
10343         2018            367  Kansas State University   
10344         2018            367  Kansas State University   
10345         2018            367  Kansas State University   
10346         2018            367  Kansas State University   
10347         2018            367  Kansas State University   
10348         2018            367  Kansas State University   
10349         2018            367  Kansas State University   
10350         2018            367  Kansas State University   
10351         2018            367  Kansas State University   
10352         2018            367  Kansas State University   
...            ...            ...                      ...   
15757         2018            367  Kansas State University   
15758         2018            367  Kansas State University   
15759         2018            367  Kansas State University   
15760         2018            367  Kansas State University   
15761         2018            367  Kansas State University   
15762         2018            367  Kansas State University   
15763         2018            367  Kansas State University   
15764         2018            367  Kansas State University   
15765         2018            367  Kansas State University   
15766         2018            367  Kansas State University   
15767         2018            367  Kansas State University   
15768         2018            367  Kansas State University   
15769         2018            367  Kansas State University   
15770         2018            367  Kansas State University   
15771         2018            367  Kansas State University   
15772         2018            367  Kansas State University   
15773         2018            367  Kansas State University   
15774         2018            367  Kansas State University   
15775         2018            367  Kansas State University   
15776         2018            367  Kansas State University   
15777         2018            367  Kansas State University   
15778         2018            367  Kansas State University   
15779         2018            367  Kansas State University   
15780         2018            367  Kansas State University   
15781         2018            367  Kansas State University   
15782         2018            367  Kansas State University   
15783         2018            367  Kansas State University   
15784         2018            367  Kansas State University   
15785         2018            367  Kansas State University   
15786         2018            367  Kansas State University   

                   Employee_Name                   Job_Title  Total_Gross_Pay  \
10323             Hao  Yangfan                       Scholar          1653.00   


Answer: See table yourself

#### Question 4: At Kansas University University, who got paid the most in the year of 2018? Is it considered an increase or decrease of compensation for that employee, compared to the year of 2017? How big is the difference?

In [42]:
merged = pd.merge(KSU, first, on="Employee_Name")

In [43]:
merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4600 entries, 0 to 4599
Data columns (total 14 columns):
Fiscal_Year          4600 non-null int64
Agency_Number_x      4600 non-null int64
Agency_Name_x        4600 non-null object
Employee_Name        4600 non-null object
Job_Title_x          4600 non-null object
Total_Gross_Pay_x    4600 non-null float64
Overtime_Pay_x       4600 non-null float64
Unnamed: 7           0 non-null float64
Agency_Number_y      4600 non-null int64
Agency_Name_y        4600 non-null object
Job_Title_y          4600 non-null object
Total_Gross_Pay_y    4600 non-null float64
Overtime_Pay_y       482 non-null float64
Unnamed: 6           0 non-null float64
dtypes: float64(6), int64(3), object(5)
memory usage: 539.1+ KB


In [50]:
merged.groupby(["Employee_Name", "Job_Title_x", "Total_Gross_Pay_y" ]).Total_Gross_Pay_x.sum().reset_index().sort_values("Total_Gross_Pay_x", ascending=False).head()

Employee_Name            Job_Title_x  Total_Gross_Pay_y  \
2863        Myers  Richard B              President          454615.35   
3993          Taylor  Gene F  Director of Athletics           60576.95   
1023  Dorhout  Peter Kenneth         Vice President          332307.63   
912   Dawson  Darren Merritt         Dean-Professor          328361.58   
1529  Gwinner  Kevin Patrick         Dean-Professor          345361.58   

      Total_Gross_Pay_x  
2863          531250.02  
3993          450000.20  
1023          327815.27  
912           323922.66  
1529          323922.66

In [51]:
454615.35-531250.02

-76634.67000000004

Answer: President Myeres was the Highest paid for both years. He got a 76634.67000000004 raise

### END OF THE TEST